torchtune/modules/early_exit_loss.py

```diff
@@ -0,0 +1,293 @@
+            should be output to calculate their losses.
+        max_steps (int): The maximum number of steps in the curriculum.
+        train_last_layer (bool, optional): Whether to always calculate loss for the last layer. Defaults to True.
+        verbose (bool, optional): Whether to print verbose logs. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        do_output_hidden_states: List[bool],
+        max_steps: int,
+        train_last_layer: bool = True,
+        verbose: bool = False,
+    ):
+        self._init_do_output_hidden_states = do_output_hidden_states
+        self._do_output_hidden_states = do_output_hidden_states
+        self.train_last_layer = train_last_layer
+        self.verbose = verbose
+        self.max_steps = max_steps
+
+    def step(self) -> None:
+        """
+        Perform a step in the curriculum. Should be called at the end of each iteration during training.
+        """
+        pass
+
+    def get(self) -> np.ndarray:
+        """
+        Get the current output hidden states.
+        Returns:
+            np.ndarray: A list indicating whether we should calculate loss for each layer.
+        """
+        do_output_hidden_states = np.copy(self._do_output_hidden_states)
+        # Ensure last layer is trained
+        if self.train_last_layer:
+            do_output_hidden_states[-1] = True
+        return do_output_hidden_states
+
+
+class RotationalEarlyExitCurriculum(EarlyExitCurriculum):
+    """
+    A rotational early exit curriculum, which rotates the layer enablement one step forward
+    at each step.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._initial_do_output_hidden_states = np.copy(self._do_output_hidden_states)
+
+    def step(self):
+        """
+        Rotate the layer enablement one step forward.
+        This method updates the `do_output_hidden_states` attribute by rotating it one position to the right.
+        """
+        # Rotate layer enablement one step forward
+        self._do_output_hidden_states = np.roll(self._do_output_hidden_states, 1)
+
+        if self.verbose:
+            log.info(
+                f"Updated self._do_output_hidden_states to {self._do_output_hidden_states}."
+            )
+
+
+class GradualEarlyExitCurriculum(EarlyExitCurriculum):
+    """
+    A gradual early exit curriculum, which gradually enables more layers (starting from the last layer) as training progresses.
+    Args:
+        *args: Positional arguments passed to the parent EarlyExitCurriculum class.
+        percent_scale (float, optional): A scaling factor to determine at which percentage
+            of steps, all the layers will be enabled. At `steps = max_steps / percent_scale`, all the layers will be enabled.
+        **kwargs: Keyword arguments passed to the parent EarlyExitCurriculum class.
+    """
+
+    def __init__(
+        self,
+        *args,
```

Comment:
Is the usage of args just for saving a few lines and not repeating the same args from the parent class? In this case I think mixing child args in the model makes it more confusing and not necessarily worth the lines saved

Reply:
Done repeating args in e5534eaf48ac8b4f42c7629ed76727d2d95e7b86